worker/rhpv3.go

```diff
@@ -376,117 +373,113 @@ func (w *worker) initTransportPool() {
+	return withAccountLock(ctx, a.as, a.id, a.host, false, func(_ api.Account) error {
+		amt, err := amtFn()
+		if err != nil {
+			return err
+		}
+		return a.as.AddBalance(ctx, a.id, a.host, amt.Big())
+	})
+}
+
+// WithSync syncs an accounts balance with the bus. To do so, the account is
+// locked while the balance is fetched through balanceFn.
+func (a *account) WithSync(ctx context.Context, balanceFn func() (types.Currency, error)) error {
+	return withAccountLock(ctx, a.as, a.id, a.host, true, func(_ api.Account) error {
+		balance, err := balanceFn()
+		if err != nil {
+			return err
+		}
+		return a.as.SetBalance(ctx, a.id, a.host, balance.Big())
+	})
 }
 
 // WithWithdrawal decreases the balance of an account by the amount returned by
 // amtFn. The amount is still withdrawn if amtFn returns an error since some
 // costs are non-refundable.
 func (a *account) WithWithdrawal(ctx context.Context, amtFn func() (types.Currency, error)) error {
-	account, lockID, err := a.as.LockAccount(ctx, a.id, a.host, false, accountLockingDuration)
-	if err != nil {
-		return err
-	}
-	defer func() {
-		unlockCtx, cancel := context.WithTimeout(a.shutdownCtx, 10*time.Second)
-		a.as.UnlockAccount(unlockCtx, a.id, lockID)
-		cancel()
-	}()
+	return withAccountLock(ctx, a.as, a.id, a.host, false, func(account api.Account) error {
+		// return early if the account needs to sync
+		if account.RequiresSync {
+			return fmt.Errorf("%w; account requires resync", errBalanceInsufficient)
+		}
 
-	// return early if the account needs to sync
-	if account.RequiresSync {
-		return fmt.Errorf("%w; account requires resync", errBalanceInsufficient)
-	}
+		// return early if our account is not funded
+		if account.Balance.Cmp(big.NewInt(0)) <= 0 {
+			return errBalanceInsufficient
+		}
 
-	// return early if our account is not funded
-	if account.Balance.Cmp(big.NewInt(0)) <= 0 {
-		return errBalanceInsufficient
-	}
+		// execute amtFn
+		amt, err := amtFn()
+		if isBalanceInsufficient(err) {
+			// in case of an insufficient balance, we schedule a sync
+			if scheduleErr := a.scheduleSync(); scheduleErr != nil {
+				err = fmt.Errorf("%w; failed to set requiresSync flag on bus, error: %v", err, scheduleErr)
+			}
+		}
 
-	// execute amtFn
-	amt, err := amtFn()
-	if isBalanceInsufficient(err) {
-		// in case of an insufficient balance, we schedule a sync
-		scheduleCtx, cancel := context.WithTimeout(a.shutdownCtx, 10*time.Second)
-		defer cancel()
-		err2 := a.as.ScheduleSync(scheduleCtx, a.id, a.host)
-		if err2 != nil {
-			err = fmt.Errorf("%w; failed to set requiresSync flag on bus, error: %v", err, err2)
+		// if an amount was returned, we withdraw it.
+		if withdrawErr := a.withdrawFromBalance(amt); withdrawErr != nil {
+			err = fmt.Errorf("%w; failed to withdraw from account, error: %v", err, withdrawErr)
 		}
-	}
 
-	// if the amount is zero, we are done
-	if amt.IsZero() {
 		return err
+	})
+}
+
+func (a *account) withdrawFromBalance(amt types.Currency) error {
```

Comment:
I'd either inline these which might be a bit nasty or go all the way and create a wrapper like so for the whole AccountStore interface:

```go
accountStore struct {
  timeout time.Duration
  store AccountStore
}

func (as *accountStore) AddBalance(id rhpv3.Account, hk types.PublicKey, amt *big.Int) error {
  ctx, cancel := context.WithTimeout(context.Background(), as.timeout)
  defer cancel()
  return as.store.AddBalance(ctx, id, hk, amt)
}
```

It's a bit confusing to have both `a.scheduleSync()` and `a.as.ScheduleSync()` but with a wrapper it will just be `a.as.ScheduleSync` and you are 100% guaranteed to never call the bus with the wrong context for any account store endpoint.

Reply:
Inlining was not so bad, wdyt? I don't really like the wrapper.

```go
// execute amtFn
amt, err := amtFn()

// in case of an insufficient balance, we schedule a sync
if isBalanceInsufficient(err) {
    ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
    err = errors.Join(err, a.as.ScheduleSync(ctx, a.id, a.host))
    cancel()
}

// if an amount was returned, we withdraw it.
if !amt.IsZero() {
    ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
    err = errors.Join(err, a.as.AddBalance(ctx, a.id, a.host, new(big.Int).Neg(amt.Big())))
    cancel()
}
```